### ISSUE TITLE
Upgrade Carbon to Spark 1.6

### ISSUE THREAD

**Issue by ravipesala** [HuaweiBigData/carbondata#618]
Following are major changes needed for carbon to upgrade to Spark 1.6.
Aggregation interfaces got completely changed in Spark 1.6. So carbon also need to update aggregate interfaces to keep the push down aggregation works. 
Current implementation in carbon-spark layer to pushdown and calculate aggregates no longer works. 
We need to come up with new approach to push down and calculate aggregates.
Distinct aggregate like distinct-count or distinct-sum cannot be pushed down to carbon layer as Spark is changed the execution plan for distinct aggregates (This plan is multi step plan to optimize calculating distinct aggregate and cannot be pushed down to carbon layer)

**Comment by jackylk** [comment]
Yes, `Aggregate` physical plan is replaced by `TunstenAggregate`. It required changes in carbon-spark for aggregation related logic. 
Also, as we are going to refactor related code, I suggest to have more options for user to set the integration level with Spark and CarbonData. For example, user should be enable to choose from following:

1. Integrate with CarbonData's basic features only in executor side, leveraging intra-file indexing in `InputFormat/OutputFormat`. All these features are enabled: multidiemsional index, inverted index, local dictionary, column group, vectorized encode/decode

2. Integrate with CarbonData's advanced features (any feature require cross file processing) in driver side, user can config to use:

- Inter-file indexing
- Computation pushdown strategies in optimizer
- Table leve dictionary for late materialization

What do you think?
